File: SpringMicroservices/microservice-course/src/main/java/com/microservice/course/controller/CourseController.java
package com.microservice.course.controller;

import com.microservice.course.entities.Course;
import com.microservice.course.http.response.StudentByCourseResponse;
import com.microservice.course.service.ICourseService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.*;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@CrossOrigin(origins = "http://localhost:3000") // 🔹 Permite todos los orígenes temporalmente
@RestController
@RequestMapping("/api/course")
public class CourseController {

    @Autowired
    private ICourseService courseService;

    // ✅ CREAR (INSERT)
    @PostMapping
    public ResponseEntity<Course> create(@Valid @RequestBody Course course) {
        Course saved = courseService.save(course);
        return new ResponseEntity<>(saved, HttpStatus.CREATED);
    }

    // ✅ LEER TODOS (READ ALL)
    @GetMapping
    public ResponseEntity<List<Course>> findAll() {
        return ResponseEntity.ok(courseService.findAll());
    }

    // ✅ LEER POR ID (READ BY ID)
    @GetMapping("/{id}")
    public ResponseEntity<?> findById(@PathVariable Long id) {
        Course course = courseService.findById(id);
        if (course == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body("Curso no encontrado");
        }
        return ResponseEntity.ok(course);
    }

    // ✅ LEER ESTUDIANTES DE UN CURSO (READ STUDENTS BY COURSE)
    @GetMapping("/{courseId}/students")
    public ResponseEntity<?> findStudentsByCourseId(@PathVariable Long courseId) {
        StudentByCourseResponse response = courseService.findStudentsByCourseId(courseId);
        if (response == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body("No se encontraron estudiantes para este curso");
        }
        return ResponseEntity.ok(response);
    }

    // ✅ ACTUALIZAR (UPDATE)
    @PutMapping("/{id}")
    public ResponseEntity<?> update(@PathVariable Long id, @RequestBody Course updatedCourse) {
        Course existing = courseService.findById(id);
        if (existing == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Curso no encontrado");
        }

        existing.setName(updatedCourse.getName());
        existing.setTeacher(updatedCourse.getTeacher());

        courseService.save(existing);
        return ResponseEntity.ok(existing);
    }

    // ✅ ELIMINAR (DELETE)
    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(@PathVariable Long id) {
        Course existing = courseService.findById(id);
        if (existing == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Curso no encontrado");
        }

        courseService.deleteById(id);
        return ResponseEntity.ok("Curso eliminado correctamente");
    }
}

File: Nest/nest/src/course/course.controller.ts
import { Controller, Get, Post, Put, Delete, Param, Body } from '@nestjs/common';
import { CourseService } from './course.service';

@Controller('api/course') // 👈 Alineado con Spring Boot
export class CourseController {
  constructor(private readonly courseService: CourseService) {}

  // ✅ Obtener todos los cursos
  @Get('all')
  getAll() {
    return this.courseService.getAllCourses();
  }

  // ✅ Buscar curso por ID
  @Get(':id')
  getById(@Param('id') id: number) {
    return this.courseService.getCourseById(id);
  }

  // ✅ Crear nuevo curso
  @Post()
  create(@Body() course: { name: string; teacher?: string }) {
    return this.courseService.createCourse(course);
  }

  // ✅ Actualizar curso existente
  @Put(':id')
  update(@Param('id') id: number, @Body() course: { name: string; teacher?: string }) {
    return this.courseService.updateCourse(id, course);
  }

  // ✅ Eliminar curso
  @Delete(':id')
  delete(@Param('id') id: number) {
    return this.courseService.deleteCourse(id);
  }

  // ✅ Obtener estudiantes por curso
  @Get(':courseId/students')
  getStudentsByCourseId(@Param('courseId') courseId: number) {
    return this.courseService.getStudentsByCourseId(courseId);
  }
}
